internal/hub/conf.go

```diff
@@ -0,0 +1,208 @@
+	// ClientID of the Cerbos Hub credential. Defaults to the value of the CERBOS_HUB_CLIENT_ID environment variable.
+	ClientID string `yaml:"clientID" conf:",example=92B0K05B6HOF"`
+	// ClientSecret of the Cerbos Hub credential. Defaults to the value of the CERBOS_HUB_CLIENT_SECRET environment variable.
+	ClientSecret string `yaml:"clientSecret" conf:",example=${CERBOS_HUB_CLIENT_SECRET}"`
+	// WorkspaceSecret used to decrypt the bundles. Defaults to the value of the CERBOS_HUB_WORKSPACE_SECRET environment variable.
+	WorkspaceSecret string `yaml:"workspaceSecret" conf:",example=${CERBOS_HUB_WORKSPACE_SECRET}"`
+	// Deprecated: Use PDPID
+	InstanceID string `yaml:"instanceID" conf:",ignore"`
+	// Deprecated: Use WorkspaceSecret
+	SecretKey string `yaml:"secretKey" conf:",ignore"`
+}
+
+func (cc *CredentialsConf) Validate() (outErr error) {
+	// SecretKey was renamed to WorkspaceSecret in Cerbos 0.31.0
+	if cc.WorkspaceSecret == "" && cc.SecretKey != "" {
+		cc.WorkspaceSecret = cc.SecretKey
+	}
+
+	// InstanceID was renamed to PDPID in Cerbos 0.31.0
+	if cc.PDPID == "" && cc.InstanceID != "" {
+		cc.PDPID = cc.InstanceID
+	}
+
+	// We don't do any validation here because some fields are optional depending on the use case.
+
+	return nil
+}
+
+func (cc *CredentialsConf) LoadFromEnv() {
+	if cc.ClientID == "" {
+		cc.ClientID = GetEnv(ClientIDKey)
+	}
+
+	if cc.ClientSecret == "" {
+		cc.ClientSecret = GetEnv(ClientSecretKey)
+	}
+
+	if cc.PDPID == "" {
+		cc.PDPID = GetEnv(PDPIDKey)
+	}
+
+	if cc.WorkspaceSecret == "" {
+		cc.WorkspaceSecret = GetEnv(WorkspaceSecretKey)
+	}
+}
+
+func (cc CredentialsConf) ToCredentials() (*credentials.Credentials, error) {
+	return credentials.New(cc.ClientID, cc.ClientSecret, cc.WorkspaceSecret)
+}
+
+// ConnectionConf holds configuration for the remote connection.
+type ConnectionConf struct {
+	// TLS defines settings for TLS connections.
+	TLS TLSConf `yaml:"tls"`
+	// APIEndpoint is the address of the API server.
+	APIEndpoint string `yaml:"apiEndpoint" conf:"required,example=https://api.cerbos.cloud"`
+	// BootstrapEndpoint is the addresses of the server serving the bootstrap configuration.
+	BootstrapEndpoint string `yaml:"bootstrapEndpoint" conf:"required,example=https://cdn.cerbos.cloud"`
+	// MinRetryWait is the minimum amount of time to wait between retries.
+	MinRetryWait time.Duration `yaml:"minRetryWait" conf:",example=1s"`
+	// MaxRetryWait is the maximum amount of time to wait between retries.
+	MaxRetryWait time.Duration `yaml:"maxRetryWait" conf:",example=120s"`
+	// NumRetries is the number of times to retry before giving up.
+	NumRetries uint `yaml:"numRetries" conf:",example=5"`
+	// HeartbeatInterval is the interval for sending regular heartbeats.
+	HeartbeatInterval time.Duration `yaml:"heartbeatInterval" conf:",example=2m"`
+}
+
+func (cc ConnectionConf) IsUnset() bool {
+	return cc == ConnectionConf{}
+}
+
+func (cc *ConnectionConf) Validate() error {
```

Comment:
Should this be a `SetDefaults` call?

Reply:
I am actually abusing the `Validate` method here. The reason it's not `SetDefaults` is because we want to prevent the user from setting illegal values and revert them to sensible values. That has to be done at the end of loading the config hence why it's in `Validate`